invenio_oaiserver/percolator.py

```diff
@@ -110,61 +113,124 @@ def _new_percolator(spec, search_pattern):
-            _create_percolator_mapping(
-                index, percolator_doc_type, mapping_path)
-            current_search_client.index(
-                index=_build_percolator_index_name(index),
-                doc_type=percolator_doc_type,
-                id='oaiset-{}'.format(spec),
-                body={'query': query}
-            )
+            try:
+                percolator_doc_type = _get_percolator_doc_type(index)
+                _create_percolator_mapping(
+                    index, percolator_doc_type, mapping_path)
+                current_search_client.index(
+                    index=_build_percolator_index_name(index),
+                    doc_type=percolator_doc_type,
+                    id='oaiset-{}'.format(spec),
+                    body={'query': query}
+                )
+            except Exception as e:
+                # caught on schemas, which do not contain the query field
+                current_app.logger.warning(e)
+                pass
 
 
 def _delete_percolator(spec, search_pattern):
-    """Delete percolator associated with the new oaiset."""
-    if spec:
-        for index in current_search.mappings.keys():
-            # Create the percolator doc_type in the existing index for >= ES5
-            percolator_doc_type = _get_percolator_doc_type(index)
-            _create_percolator_mapping(index, percolator_doc_type)
-            current_search_client.delete(
-                index=_build_percolator_index_name(index),
-                doc_type=percolator_doc_type,
-                id='oaiset-{}'.format(spec), ignore=[404]
-            )
-
-
-def _build_cache():
-    """Build sets cache."""
-    sets = current_oaiserver.sets
-    if sets is None:
-        # build sets cache
-        sets = current_oaiserver.sets = [
-            oaiset.spec for oaiset in OAISet.query.filter(
-                OAISet.search_pattern.is_(None)).all()]
-    return sets
-
-
-def get_record_sets(record):
-    """Find matching sets."""
-    # get lists of sets with search_pattern equals to None but already in the
-    # set list inside the record
-    record_sets = set(current_oaiserver.record_sets_fetcher(record))
-    for spec in _build_cache():
-        if spec in record_sets:
-            yield spec
-
-    # get list of sets that match using percolator
-    index, doc_type = RecordIndexer().record_to_index(record)
-    document = record.dumps()
-    percolator_doc_type = _get_percolator_doc_type(index)
-    _create_percolator_mapping(index, percolator_doc_type)
-    results = _percolate_query(index, doc_type, percolator_doc_type, document)
+    """Delete percolator associated with the removed/updated oaiset."""
+    # Create the percolator doc_type in the existing index for >= ES5
+    for index, mapping_path in current_search.mappings.items():
+        percolator_doc_type = _get_percolator_doc_type(index)
+        _create_percolator_mapping(index, percolator_doc_type)
```

Comment:
Why do we `create` again? the inner code of the function basically creates it if doesnt exists... seems to simply be there to avoid a potential "cannot delete what's not there" error... but that should be taken care of by the `404` below.